### PR TITLE
feat: hold expense in one transaction view

### PR DESCRIPTION
**Pull request by tienifr**
### Details

This PR adds the ability to hold expense in one transaction view.

### Fixed Issues

$ https://github.com/Expensify/App/issues/39989
PROPOSAL: https://github.com/Expensify/App/issues/39989#issuecomment-2104969004

### Tests

1. Submit an expense (make sure it's the only expense in the expense report)
2. Press the report preview to open one transaction view
3. Put the expense on hold
4. Verify the hold education interstitial appears
5. Verify the hold message and reason appear in the expense report
6. Verify `Hold` banner appears in report header
7. Unhold the expense
8. Verify the unhold message appears
9. Verify `Hold` banner disappears in report header

- [x] Verify that no errors appear in the JS console

### Offline tests

NA

### QA Steps

1. Submit an expense (make sure it's the only expense in the expense report)
2. Press the report preview to open one transaction view
3. Put the expense on hold
4. Verify the hold education interstitial appears
5. Verify the hold message and reason appear in the expense report
6. Verify `Hold` banner appears in report header
7. Unhold the expense
8. Verify the unhold message appears
9. Verify `Hold` banner disappears in report header

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

image

image

image

image

image

</details>

<details>
<summary>Android: mWeb Chrome</summary>

image

image

image

image

</details>

<details>
<summary>iOS: Native</summary>

Simulator Screenshot - iPhone 15 Pro Max - 2024-05-21 at 14 13 50

Simulator Screenshot - iPhone 15 Pro Max - 2024-05-21 at 14 13 54

Simulator Screenshot - iPhone 15 Pro Max - 2024-05-21 at 14 14 08

</details>

<details>
<summary>iOS: mWeb Safari</summary>

Simulator Screenshot - iPhone 15 Pro Max - 2024-05-21 at 14 18 48
Simulator Screenshot - iPhone 15 Pro Max - 2024-05-21 at 14 19 12
Simulator Screenshot - iPhone 15 Pro Max - 2024-05-21 at 14 19 15

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/113963320/64b05117-c3f8-4dd6-82ca-9e3f107fe3c5

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/113963320/8eaa7d1b-7a63-44f4-83fa-6471b74be149

</details>

### CODE REVIEW

**Comment by shawnborton** [comment]
Looks pretty good to me from a design standpoint. cc @trjExpensify @JmillsExpensify to confirm accuracy.

**Comment by trjExpensify** [comment]
Nice same, looks good!

**Review by dukenv0307 — APPROVED** [review]

File: src/components/MoneyReportHeader.tsx
Comment (NikkiWines) [inline src/components/MoneyReportHeader.tsx:41]: NAB: 
```suggestion
    /** All the data for the transaction in the one transaction view */
```

File: src/components/MoneyReportHeader.tsx
Comment (NikkiWines) [inline src/components/MoneyReportHeader.tsx:127]: Did the linter complain about this because you're passing `transaction` to the component now? Ideally, we'd use a non-abbreviated variable name here but if it's because of that then I guess it's ok.

**Review by NikkiWines — CHANGES_REQUESTED** [review]
Looks good, though the overall height of the report header changes when going between an open request with `Next Steps` and a held request. 

https://github.com/Expensify/App/assets/16747822/769f3db8-893e-4394-8dee-d84ec2e0e2f4

**Comment by dubielzyk-expensify** [comment]
Looks good to me, but I think the button is too big, no? Looking at this issue I think we're using `size: medium` for the button but this looks `large` to me?

**Comment by shawnborton** [comment]
I think you are right - I updated the mock above to reflect the medium/default size.

File: src/components/MoneyReportHeader.tsx
Comment (luacmartins) [inline src/components/MoneyReportHeader.tsx:90]: NAB this name is a bit misleading because `isMoneyRequestReport` also returns true for IOU reports which doesn't have approvals.

File: src/components/MoneyReportHeader.tsx
Comment (luacmartins) [inline src/components/MoneyReportHeader.tsx:102]:
```suggestion
    const transactionIDs = TransactionUtils.getAllReportTransactions(moneyRequestReport?.reportID).map((transaction) => transaction.transactionID);
```

File: src/components/MoneyReportHeader.tsx
Comment (luacmartins) [inline src/components/MoneyReportHeader.tsx:187]:
```suggestion
        const transactionID = requestParentReportAction?.actionName === CONST.REPORT.ACTIONS.TYPE.IOU ? requestParentReportAction.originalMessage?.IOUTransactionID ?? '' : '';
```

File: src/components/MoneyReportHeader.tsx
Comment (luacmartins) [inline src/components/MoneyReportHeader.tsx:190]:
```suggestion
            IOU.unholdRequest(transactionID, transactionThreadReport.reportID);
```

File: src/components/MoneyReportHeader.tsx
Comment (luacmartins) [inline src/components/MoneyReportHeader.tsx:193]:
```suggestion
            Navigation.navigate(ROUTES.MONEY_REQUEST_HOLD_REASON.getRoute(policy?.type ?? CONST.POLICY.TYPE.PERSONAL, transactionID, transactionThreadReport.reportID, activeRoute));
```

File: src/components/MoneyReportHeader.tsx
Comment (luacmartins) [inline src/components/MoneyReportHeader.tsx:225]: Or we can just reuse `ReportUtils.isIOUReport`
```suggestion
        const isRequestIOU = chatReport?.type === CONST.REPORT.TYPE.IOU;
```

**Review by luacmartins — CHANGES_REQUESTED** [review]

**Comment by dannymcclain** [comment]
@tienifr so are you saying that Hold and Next steps were basically using different padding, which is what led to the header height changing?

Definitely want to wait for @shawnborton to weigh in, but your screenshots are looking pretty good to me.

**Review by luacmartins — DISMISSED** [review]
LGTM

**Comment by luacmartins** [comment]
@NikkiWines @shawnborton let me know if you'd like to review again.

**Review by NikkiWines — COMMENTED** [review]
Retested and works great but you've got conflicts now @tienifr

**Review by NikkiWines — COMMENTED** [review]
Changes after the merge look good but now there seems to be somewhat of a delay when upholding the expense 🤔  

Screen-recording is also showing a weird border around the HOLD banner but it looks good to me irl ¯\_(ツ)_/¯  

https://github.com/Expensify/App/assets/16747822/f29e95af-0fec-4c09-939d-56965ed74630

cc: @luacmartins do you also see the delay for the upholding?

**Comment by luacmartins** [comment]
Seems like @tienifr is able to reproduce the issue. I'll wait for them to push changes before reviewing again

File: src/components/MoneyReportHeader.tsx
Comment (luacmartins) [inline src/components/MoneyReportHeader.tsx:69]: Connecting to all transactions seems overkill. Can we use a selector to get only the specific report transactions? I believe that we'd also be able to remove the call to `TransactionUtils.getAllReportTransactions` in `hasOnlyHeldExpenses` if we only pass report transactions to it

**Review by luacmartins — COMMENTED** [review]
Still waiting on @tienifr to address the comments above

**Comment by trjExpensify** [comment]
@tienifr there's conflicts to resolve now, and a comments above to address. Can you give us an ETA for completing this, please?

**Comment by NikkiWines** [comment]
@tienifr there are linter errors btw

**Comment by trjExpensify** [comment]
@tienifr did you address the comments?

**Comment by luacmartins** [comment]
It seems like this comment is still pending @tienifr https://github.com/Expensify/App/pull/42374#discussion_r1620762402

File: src/components/MoneyReportHeader.tsx
Comment (luacmartins) [reply src/components/MoneyReportHeader.tsx:69]: Thanks for the clarification. I agree that we should address the issue with `allTransactions` in `TransactionUtils` since that won't be up to date for components using it. Given that subscribing to all transactions doesn't increase the number of re-renders, I'm fine with the current solution.

**Review by luacmartins — APPROVED** [review]

**Review by NikkiWines — APPROVED** [review]
:shipit:

**Comment by luacmartins** [comment]
@tienifr please work on a fix for the issues encountered and submit a new PR

**Pull request merged**
